File: src/main/java/com/abysalto/middleware/service/ProductService.java
package com.abysalto.middleware.service;

import com.abysalto.middleware.domain.Product;
import com.abysalto.middleware.domain.ProductPage;
import com.abysalto.middleware.dto.PagedResponse;
import com.abysalto.middleware.dto.ProductDetailDto;
import com.abysalto.middleware.dto.ProductSummaryDto;
import com.abysalto.middleware.source.ProductSource;
import org.springframework.stereotype.Service;

import java.math.BigDecimal;
import java.util.List;

/**
 * Application service over the {@link ProductSource} abstraction. Responsible for pagination,
 * mapping to API DTOs and the in-service price-range filtering that the upstream cannot do.
 *
 * <p>Cache-backed upstream access for search/filter is delegated to {@link ProductQueryCache} (a
 * separate bean so Spring's cache proxy is honoured and the price-filter candidate set can be cached
 * independently of pagination); this service handles orchestration, in-service paging and mapping.
 */
@Service
public class ProductService {

    private final ProductSource source;
    private final ProductQueryCache queries;
    private final ProductMapper mapper;

    public ProductService(ProductSource source, ProductQueryCache queries, ProductMapper mapper) {
        this.source = source;
        this.queries = queries;
        this.mapper = mapper;
    }

    /** Paginated trimmed list of all products. */
    public PagedResponse<ProductSummaryDto> list(int page, int size) {
        ProductPage result = source.list(offset(page, size), size);
        return toSummaryPage(result.items(), page, size, result.total());
    }

    /** Full detail of a single product. */
    public ProductDetailDto getById(long id) {
        return mapper.toDetail(source.getById(id));
    }

    /**
     * Filters by category and/or price range (combinable). Category is pushed down to the source;
     * price range is applied in-service.
     *
     * <p>When no price bound is present the source paginates directly (cached per page). When a price
     * bound is present the full price-filtered candidate set is fetched once (cached by category and
     * price bounds via {@link ProductQueryCache}, independent of pagination) and the requested page is
     * sliced from it here, so paging through the result never re-fetches the upstream catalog.
     */
    public PagedResponse<ProductSummaryDto> filter(String category, BigDecimal minPrice, BigDecimal maxPrice,
                                                   int page, int size) {
        boolean hasPriceFilter = minPrice != null || maxPrice != null;

        if (!hasPriceFilter) {
            ProductPage result = queries.categoryPage(category, page, size);
            return toSummaryPage(result.items(), page, size, result.total());
        }

        List<Product> filtered = queries.priceFilteredCandidates(category, minPrice, maxPrice);
        List<Product> pageItems = paginate(filtered, page, size);
        return toSummaryPage(pageItems, page, size, filtered.size());
    }

    /** Free-text search by product name (pushed down to the source, cached per page). */
    public PagedResponse<ProductSummaryDto> searchByName(String query, int page, int size) {
        ProductPage result = queries.search(query, page, size);
        return toSummaryPage(result.items(), page, size, result.total());
    }

    /** Available category identifiers. */
    public List<String> categories() {
        return source.categories();
    }

    // --- helpers -----------------------------------------------------------

    private static int offset(int page, int size) {
        return page * size;
    }

    private static <T> List<T> paginate(List<T> items, int page, int size) {
        int from = Math.min(offset(page, size), items.size());
        int to = Math.min(from + size, items.size());
        return items.subList(from, to);
    }

    private PagedResponse<ProductSummaryDto> toSummaryPage(List<Product> products, int page, int size, long total) {
        List<ProductSummaryDto> summaries = products.stream().map(mapper::toSummary).toList();
        return PagedResponse.of(summaries, page, size, total);
    }
}

File: src/Middleware.Core/Services/ProductService.cs
using Middleware.Core.Abstractions;
using Middleware.Core.Domain;
using Middleware.Core.Dtos;

namespace Middleware.Core.Services;

/// <summary>
/// Application service over the <see cref="IProductSource"/> abstraction. Responsible for pagination,
/// mapping to API DTOs and the in-service price-range filtering that the upstream cannot do.
///
/// <para>Cache-backed upstream access for search/filter is delegated to <see cref="ProductQueryCache"/>
/// (a separate class so the single-flight cache is honoured and the price-filter candidate set can be
/// cached independently of pagination); this service handles orchestration, in-service paging and mapping.</para>
/// </summary>
public sealed class ProductService(IProductSource source, IProductQueryCache queries, ProductMapper mapper)
{
    /// <summary>Paginated trimmed list of all products.</summary>
    public async Task<PagedResponse<ProductSummaryDto>> ListAsync(int page, int size, CancellationToken ct = default)
    {
        var result = await source.ListAsync(Offset(page, size), size, ct);
        return ToSummaryPage(result.Items, page, size, result.Total);
    }

    /// <summary>Full detail of a single product.</summary>
    public async Task<ProductDetailDto> GetByIdAsync(long id, CancellationToken ct = default)
    {
        var product = await source.GetByIdAsync(id, ct);
        return mapper.ToDetail(product);
    }

    /// <summary>
    /// Filters by category and/or price range (combinable). Category is pushed down to the source;
    /// price range is applied in-service.
    ///
    /// <para>When no price bound is present the source paginates directly (cached per page). When a price
    /// bound is present the full price-filtered candidate set is fetched once (cached by category and
    /// price bounds via <see cref="ProductQueryCache"/>, independent of pagination) and the requested
    /// page is sliced from it here, so paging through the result never re-fetches the upstream catalog.</para>
    /// </summary>
    public async Task<PagedResponse<ProductSummaryDto>> FilterAsync(
        string? category, decimal? minPrice, decimal? maxPrice, int page, int size, CancellationToken ct = default)
    {
        var hasPriceFilter = minPrice is not null || maxPrice is not null;

        if (!hasPriceFilter)
        {
            var result = await queries.CategoryPageAsync(category, page, size, ct);
            return ToSummaryPage(result.Items, page, size, result.Total);
        }

        var filtered = await queries.PriceFilteredCandidatesAsync(category, minPrice, maxPrice, ct);
        var pageItems = Paginate(filtered, page, size);
        return ToSummaryPage(pageItems, page, size, filtered.Count);
    }

    /// <summary>Free-text search by product name (pushed down to the source, cached per page).</summary>
    public async Task<PagedResponse<ProductSummaryDto>> SearchByNameAsync(
        string? query, int page, int size, CancellationToken ct = default)
    {
        var result = await queries.SearchAsync(query, page, size, ct);
        return ToSummaryPage(result.Items, page, size, result.Total);
    }

    /// <summary>Available category identifiers.</summary>
    public Task<IReadOnlyList<string>> CategoriesAsync(CancellationToken ct = default) => source.CategoriesAsync(ct);

    // --- helpers -----------------------------------------------------------

    private static int Offset(int page, int size) => page * size;

    private static IReadOnlyList<Product> Paginate(IReadOnlyList<Product> items, int page, int size)
    {
        var from = Math.Min(Offset(page, size), items.Count);
        var to = Math.Min(from + size, items.Count);
        return items is List<Product> list ? list.GetRange(from, to - from) : items.Skip(from).Take(to - from).ToList();
    }

    private PagedResponse<ProductSummaryDto> ToSummaryPage(IReadOnlyList<Product> products, int page, int size, long total)
    {
        var summaries = products.Select(mapper.ToSummary).ToList();
        return PagedResponse<ProductSummaryDto>.Of(summaries, page, size, total);
    }
}
